[analyzer] Allow recursive functions to be trivial.

rniwa opened the pull request.

Commit:
[analyzer] Allow recursive functions to be trivial.

Commit:
Fix formatting.

MitalAshok commented:
You should add a test for mutually recursive functions. I suspect something like this doesn't work:

```c++
    int non_trivial();
    int f(bool b) { return g(!b) + non_trivial(); }
    int g(bool b) { return b ? f(b) : 1; }
    
getFieldTrivial().f(true);  // expected-warning {{...}}
getFieldTrivial().g(true);  // expected-warning {{...}}
```

Since when analyzing `f`, `f` enters the cache as trivial, `g` is analyzed and sees `f` in the cache as trivial, so `g` is marked trivial, but then `f` is marked as non-trivial (so `g` should have been marked as non-trivial, but is marked trivial).

If that is an issue, one way to do this "properly" would be to create a graph, and a function is non-trivial if it has a path to a non-trivial function (this might be too expensive to implement directly). Or we could special case simple recursion only, by changing `TrivialFunctionAnalysisVisitor` to not call `isTrivialImpl` for the current function only.

rniwa commented:
> You should add a test for mutually recursive functions. I suspect something like this doesn't work:
> 
> ```c++
>     int non_trivial();
>     int f(bool b) { return g(!b) + non_trivial(); }
>     int g(bool b) { return b ? f(b) : 1; }
>     
> getFieldTrivial().f(true);  // expected-warning {{...}}
> getFieldTrivial().g(true);  // expected-warning {{...}}
> ```

Duh, of course. Thanks for pointing out the obvious flaw.

Commit:
Fix the TrivialFunctionAnalysis::isTrivialImpl for mutually recursive functions.

Instead of assuming every function to be initially trivial, we explicitly track
the set of functions that we're currently visting. When one of the currently visited
function is determined to be not trivial, we clear this set to signal that all
mutually recursive functions are non-trivial. We conclude that a function is trivial
when Visit() call on the function body returned true **AND** the set still contains
the function.

To implement this new algorithm, a new public function, IsFunctionTrivial,
is introduced to TrivialFunctionAnalysisVisitor, and various Visit functions in
TrivialFunctionAnalysisVisitor has been updated to use this function instead of
TrivialFunctionAnalysis::isTrivialImpl, which is now a wrapper for the function.

Commit:
Fix a bug that IsFunctionTrivial was not updating the cache upon completion.

Also fix a bug that IsFunctionTrivial would do a redundant traversal when
a function had been determined to be non-trivial because it's indistinguishable
if a given function had not been traversed or had been found to be non-trivial
because we were using the absense of the function in the hash set to indicate
the non-triviality of a function. Use a hash map of a function to a boolean
instead to explicitly track whether a given function had been visited or not,
and whether a given function had been determined to be non-trivial or not.

Commit:
Fix the cache updating logic.

haoNoQ commented on clang/test/Analysis/Checkers/WebKit/uncounted-obj-arg.cpp:
```diff
@@ -231,6 +231,15 @@ class RefCounted {
+  unsigned mutuallyRecursiveFunction2(int n) { return n < 0 ? 1 : (n % 3 ? mutuallyRecursiveFunction2(n - 3) : mutuallyRecursiveFunction1(n - 2)); }
+  unsigned mutuallyRecursiveFunction3(int n) { return n < 0 ? 1 : (n % 5 ? mutuallyRecursiveFunction3(n - 5) : mutuallyRecursiveFunction4(n - 3)); }
+  unsigned mutuallyRecursiveFunction4(int n) { return n < 0 ? 1 : (n % 7 ? otherFunction() : mutuallyRecursiveFunction3(n - 3)); }
+  unsigned mutuallyRecursiveFunction5(unsigned n) { return n > 100 ? 2 : (n % 2 ? mutuallyRecursiveFunction5(n + 1) : mutuallyRecursiveFunction6(n + 2)); }
```
These probably shouldn't be called "mutually" recursive because you never really go back from 6 to 5 or from 7 to 6. Though, I don't have a better name for such contraption.

haoNoQ replied:
I think it may be a good idea to add a direct test for @MitalAshok's example.
```c++
void foo() { bar(); something_non_trivial(); }
void bar() { foo(); }

void test() {
  foo();
  bar();
}
```
where none of the functions call themselves directly. I'm still not quite sure whether we're correctly confirming that `bar()` is non-trivial here, when we start the analysis from `foo()`.

rniwa replied:
Sure, added.

haoNoQ reviewed (commented):
> If that is an issue, one way to do this "properly" would be to create a graph

Note that we do have a [`CallGraph`](https://clang.llvm.org/doxygen/classclang_1_1CallGraph.html) class. The static analyzer uses it to identify top-level entry points. It doesn't look like it has a readily available solution for identifying recursive functions. But it may be a good idea to simply use the class and implement it there.

Commit:
Add an actually mutually recursive function, and rename exiting recursive functions.

haoNoQ reviewed (approved):
Ok LGTM then!

rniwa commented:
Thanks for the review!
